### PR TITLE
Update the CustomMarshallerAttributeFixer to support adding missing methods for stateless marshaller shapes

### PR DESCRIPTION
**Pull request by jkoritzinsky**
One test is disabled against a bug in the Roslyn SDK that I found.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs [hunk 1]
```diff
@@ -23,50 +28,83 @@ public class CustomMarshallerAttributeFixer : CodeFixProvider
...
+                {
+                    Document doc = fixAllContext.Solution.GetDocument(diagnostic.Location.SourceTree);
+                    SemanticModel model = await doc.GetSemanticModelAsync(fixAllContext.CancellationToken).ConfigureAwait(false);
 
-                DocumentEditor editor = await DocumentEditor.CreateAsync(document, fixAllContext.CancellationToken).ConfigureAwait(false);
+                    var entryPointTypeSymbol = (INamedTypeSymbol)model.GetEnclosingSymbol(diagnostic.Location.SourceSpan.Start, fixAllContext.CancellationToken);
+                    ITypeSymbol? managedType = GetManagedTypeInAttributeSyntax(diagnostic.Location, entryPointTypeSymbol);
 
-                switch (fixAllContext.CodeActionEquivalenceKey)
+                    SyntaxNode node = (await diagnostic.Location.SourceTree.GetRootAsync(fixAllContext.CancellationToken).ConfigureAwait(false)).FindNode(diagnostic.Location.SourceSpan);
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs:47]: Can we pull out the `await` statement into a separate line. There is a lot going on here.

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs [hunk 2]
```diff
@@ -23,50 +28,83 @@ public class CustomMarshallerAttributeFixer : CodeFixProvider
...
+                    var marshallerType = (INamedTypeSymbol)model.GetSymbolInfo(node, fixAllContext.CancellationToken).Symbol;
+                    var uniqueMarshallerFixKey = (marshallerType, managedType, ManualTypeMarshallingHelper.IsLinearCollectionEntryPoint(entryPointTypeSymbol));
+                    if (uniqueMarshallersToFix.TryGetValue(uniqueMarshallerFixKey, out List<string> membersToAdd))
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs:50]: Can we collapse this logic so getting the list is done independently of adding to it? I think we can update this so we get the list and then in one location always add to the list.

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs [hunk 3]
```diff
@@ -23,50 +28,83 @@ public class CustomMarshallerAttributeFixer : CodeFixProvider
...
+            public static FixAllProvider Instance { get; } = new CustomFixAllProvider();
+
+            public override async Task<CodeAction?> GetFixAsync(FixAllContext fixAllContext)
             {
-                SyntaxNode? root = await document.GetSyntaxRootAsync(fixAllContext.CancellationToken).ConfigureAwait(false);
-                if (root == null)
-                    return document;
+                ImmutableArray<Diagnostic> diagnostics = await GetAllDiagnosticsInScope(fixAllContext).ConfigureAwait(false);
+                Dictionary<(INamedTypeSymbol marshallerType, ITypeSymbol managedType, bool isLinearCollectionMarshaller), List<string>> uniqueMarshallersToFix = new();
+                foreach (Diagnostic diagnostic in diagnostics)
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs:39]:
```suggestion
                // Organize all the diagnostics by marshaler, managed type, and if a collection.
                foreach (Diagnostic diagnostic in diagnostics)
```

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs [hunk 4]
```diff
@@ -23,50 +28,83 @@ public class CustomMarshallerAttributeFixer : CodeFixProvider
...
+                    {
+                        uniqueMarshallersToFix.Add(uniqueMarshallerFixKey, new List<string>(diagnostic.Properties[MissingMemberNames.Key].Split(MissingMemberNames.Delimiter)));
+                    }
+                }
+
+                Dictionary<INamedTypeSymbol, INamedTypeSymbol> partiallyUpdatedSymbols = new(SymbolEqualityComparer.Default);
+
+                SymbolEditor symbolEditor = SymbolEditor.Create(fixAllContext.Solution);
+
+                foreach (var marshallerInfo in uniqueMarshallersToFix)
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs:64]:
```suggestion
                // Apply each fix
                foreach (var marshallerInfo in uniqueMarshallersToFix)
```

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs [hunk 5]
```diff
@@ -93,42 +131,267 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                    if (diagnostic.Properties.TryGetValue(MissingMemberNames.Key, out string missingMembers))
                     {
-                        missingMemberNames.AddRange(missingMembers.Split(CustomMarshallerAttributeAnalyzer.MissingMemberNames.Delimiter));
+                        missingMemberNames.UnionWith(missingMembers.Split(MissingMemberNames.Delimiter));
                     }
                 }
             }
 
-            return (missingMemberNames, requiredShapeDiagnostics);
+            return (new HashSet<string>(missingMemberNames), requiredShapeDiagnostics);
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs:150]: Why do we need to create yet another `HashSet<string>`?

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs [hunk 6]
```diff
@@ -93,42 +131,267 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
...
+            var model = await doc.GetSemanticModelAsync(ct).ConfigureAwait(false);
+
+            var entryPointTypeSymbol = (INamedTypeSymbol)model.GetEnclosingSymbol(node.SpanStart, ct);
+
+            // TODO: Convert to use the IOperation tree once IAttributeOperation is available
+            var managedTypeSymbolInAttribute = GetManagedTypeInAttributeSyntax(node.GetLocation(), entryPointTypeSymbol);
+
+            bool isLinearCollectionMarshaller = ManualTypeMarshallingHelper.IsLinearCollectionEntryPoint(entryPointTypeSymbol);
+
+            ManualTypeMarshallingHelper.TryResolveManagedType(entryPointTypeSymbol, ManualTypeMarshallingHelper.ReplaceGenericPlaceholderInType(managedTypeSymbolInAttribute, entryPointTypeSymbol, model.Compilation), isLinearCollectionMarshaller, (_, _) => { }, out ITypeSymbol managedType);
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs:164]: Please pull out `(_, _) => { }` as something that is named. It isn't clear to me why this exists or what is supposed to do.

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs [hunk 7]
```diff
@@ -93,42 +131,267 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
...
+            INamedTypeSymbol marshallerType = (INamedTypeSymbol)model.GetSymbolInfo(node, ct).Symbol;
 
-            return editor.GetChangedDocument();
+            await editor.EditOneDeclarationAsync(marshallerType, (editor, decl) => AddMissingMembers(editor, decl, marshallerType, managedType, missingMemberNames, isLinearCollectionMarshaller), ct).ConfigureAwait(false);
+
+            return editor.ChangedSolution;
+        }
+
+        private static ITypeSymbol? GetManagedTypeInAttributeSyntax(Location locationInAttribute, INamedTypeSymbol? attributedTypeSymbol)
+            => (ITypeSymbol)attributedTypeSymbol.GetAttributes().First(attr =>
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs:176]: Isn't `attributedTypeSymbol` nullable? Was this checked somewhere?

Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs:176]: I need to figure out what's going on here in our infra. I feel like the nullable warnings have been suppressed for some reason. I've been getting weird stuff like this for a bit (where obvious places for warnings aren't showing up)

Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/CustomMarshallerAttributeFixer.cs:187]: I haven't had a chance to investigate what's going on with our infrastructure. In the meantime, I've updated the nullability annotations here based on usage to be more accurate and added a comment explaining why we don't need to worry about nulls here.

File: src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/CustomMarshallerAttributeFixerTests_StatelessValueShapeValidation.cs [hunk 8]
```diff
@@ -34,13 +34,52 @@ static class MarshallerType
...
 
-            await VerifyCS.VerifyAnalyzerAsync(
+            string fixedSource = """
+                using System.Runtime.InteropServices.Marshalling;
+
+                class ManagedType {}
+                
+                [CustomMarshaller(typeof(ManagedType), MarshalMode.ManagedToUnmanagedIn, typeof(MarshallerType))]
+                [CustomMarshaller(typeof(ManagedType), MarshalMode.UnmanagedToManagedOut, typeof(MarshallerType))]
+                [CustomMarshaller(typeof(ManagedType), MarshalMode.ElementIn, typeof(MarshallerType))]
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/CustomMarshallerAttributeFixerTests_StatelessValueShapeValidation.cs:44]: Do we want a test that has one marshaller that handles more than one managed type? Wouldn't expect it to be common, but the generator/analyzer/fixer does handle that.

**Review by elinor-fung — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
TImeouts are a known issue.

**Pull request merged**